File: server/java/metadata/src/main/java/com/metaobjects/source/SourceResolution.java
package com.metaobjects.source;

import com.metaobjects.MetaDataException;
import com.metaobjects.object.MetaObject;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * THE {@code @role: primary} source lookup for the JVM ports — and the one place the
 * primary-source DIVERGENCE refusal lives, so that every caller inherits it.
 *
 * <p>It lives in the {@code metadata} module, not under a codegen module, because the
 * RUNTIME calls it too: {@link MetaObject#findPrimaryWritableSource()} feeds
 * {@code getPrimaryRdbTableName()}, which is how OMDB
 * ({@code SimpleMappingHandlerDB.getTableRef}) decides which physical table to read and
 * write. Under a codegen module the runtime would have to either invert its layering to
 * reach it or keep its own copy — and its own copy is exactly the shape this class exists
 * to forbid. Mirrors Python's {@code metaobjects.source_resolution}, whose codegen,
 * api-docs and runtime callers all inherit the refusal for free.</p>
 *
 * <h2>The refusal</h2>
 *
 * <p>An object whose {@code @role: primary} sources resolve to MORE THAN ONE physical
 * name has no single answer to give, so {@link #refuseDivergentPrimaries} throws rather
 * than picking one. The shape loads with ZERO errors: {@code ValidateOnePrimarySource}
 * enforces "exactly one primary" over OWN children only, and effective-children shadowing
 * matches an own child over a super child only on a {@code (type, name)} pair — so two
 * {@code source.rdb} nodes with DIFFERENT explicit names at two levels of an
 * {@code extends} chain never collide, and both survive the resolving source walk.</p>
 *
 * <p>Every consumer downstream binds ONE name unconditionally, with no per-site equality
 * guard. The refusal used to live in each codegen module's own {@code resolveObjectNames}
 * — which runs only when the {@code names} generator is in the run — so a
 * {@code mvn metaobjects:generate} without it, and every OMDB read and write, silently
 * bound the PARENT's table while the child declared its own. A refusal that depends on
 * which consumer asked is not a refusal.</p>
 *
 * <p>DIRECTION-BLIND: it compares every primary against every other, so it does not matter
 * which of them is writable nor which was declared first. Comparing against the first
 * primary WRITABLE source can only see a divergence when one of the two is read-only —
 * and, since {@link MetaObject#getSources(boolean)} concatenates OWN sources first and
 * only then walks the super chain, only when the read-only one is the OWN one. That is the
 * REVERSE of TypeScript and C#, whose effective-children walk starts from the super's list
 * and appends own — so the same comparison was blind to opposite shapes per port, and on
 * the JVM it was blind to both shapes the tests now pin (an own writable table beside an
 * inherited read-only view, and two writable primaries). Do not restate this ordering from
 * another port's copy of this doc; the two families genuinely differ.</p>
 *
 * <p>Two primaries AGREEING on a name is not a divergence and stays legal: the invariant
 * is that an object has ONE physical name, not that it declares one source. A read-only
 * primary beside a non-primary REPLICA does not reach it either — a replica is not
 * {@code role == primary}.</p>
 */
public final class SourceResolution {

    private SourceResolution() {}

    /**
     * Every {@code @role: primary} source of {@code obj}, RESOLVING through the
     * {@code extends} super chain (ADR-0039) — an entity inheriting its {@code source.rdb}
     * from an abstract base must see it, or it would wrongly read as unpersisted.
     *
     * <p>Private: nothing outside this class needs the LIST. The two public entry points
     * below are the whole surface — a caller either wants the resolved primary or only
     * wants the refusal, and no third shape has a use.</p>
     *
     * @param obj the object to inspect
     * @return the primary sources, in resolving order (own first, then the super chain)
     */
    private static List<MetaSource> primaryRdbSources(MetaObject obj) {
        List<MetaSource> primaries = new ArrayList<>();
        for (MetaSource src : obj.getSources(true)) {
            if (MetaSource.ROLE_PRIMARY.equals(src.getRole())) primaries.add(src);
        }
        return primaries;
    }

    /**
     * The role-scoped PRIMARY source of {@code obj}, or {@code null} when it has none
     * (#248: participation in the database derives from a declared primary source, never
     * from the object subtype). Runs {@link #refuseDivergentPrimaries} first, so every
     * caller inherits the refusal.
     *
     * @param obj the object to resolve
     * @return the primary source, or {@code null}
     */
    public static MetaSource primaryRdbSource(MetaObject obj) {
        List<MetaSource> primaries = primaryRdbSources(obj);
        refuseDivergentPrimaries(obj, primaries);
        return primaries.isEmpty() ? null : primaries.get(0);
    }

    /**
     * Refuse {@code obj} when its {@code @role: primary} sources disagree on a physical
     * name. See the class doc for the reachability analysis and for why the check is
     * direction-blind.
     *
     * @param obj the object to check
     * @throws MetaDataException when two or more primaries resolve different physical names
     */
    public static void refuseDivergentPrimaries(MetaObject obj) {
        refuseDivergentPrimaries(obj, primaryRdbSources(obj));
    }

    /**
     * The refusal itself, over a primary list the caller already has. ONE implementation
     * and ONE copy of the message: a check written twice is a check that can disagree with
     * itself, which is the same defect one level down from the one this class prevents (a
     * NAME resolved twice). The message is a cross-port contract string — the other four
     * ports carry it verbatim.
     */
    private static void refuseDivergentPrimaries(MetaObject obj, List<MetaSource> primaries) {
        List<String> distinct = primaries.stream()
            .map(MetaSource::getPhysicalName).distinct().sorted().toList();
        if (distinct.size() <= 1) return;
        // Sorted, so the message is identical in every port regardless of source order.
        String joined = distinct.stream().map(n -> "\"" + n + "\"").collect(Collectors.joining(", "));
        throw new MetaDataException(
            obj.getName() + ": role=primary sources disagree on the object's physical name — "
                + joined + ". Every consumer binds ONE name. Give them matching physical "
                + "names, or drop the extra role=primary declaration.");
    }
}

File: server/csharp/MetaObjects/Meta/SourceResolution.cs
// SourceResolution — THE @role: primary source lookup, and the one place the
// primary-source DIVERGENCE refusal lives.
//
// Ported alongside typescript/packages/metadata/src/naming.ts (primaryRdbSource),
// python/src/metaobjects/source_resolution.py and
// java/metadata/.../source/SourceResolution.java.

namespace MetaObjects.Meta;

/// <summary>
/// THE <c>@role: primary</c> source lookup — and the one place the primary-source
/// DIVERGENCE refusal lives, so that every caller inherits it.
///
/// <para>It lives in the core <c>MetaObjects</c> assembly, not in
/// <c>MetaObjects.Codegen</c>'s naming helper, because the RUNTIME asks the same
/// question: <c>M2MResolver.TableOf</c> reads <see cref="MetaObject.DbTable"/> and
/// executes SQL against the answer, going through no generator at all. Mirrors Python's
/// <c>metaobjects.source_resolution</c>, whose codegen, api-docs and runtime callers all
/// inherit the refusal for free.</para>
///
/// <para><b>The refusal.</b></para>
///
/// <para>An object whose <c>@role: primary</c> sources resolve to MORE THAN ONE physical
/// name has no single answer to give, so <see cref="RefuseDivergentPrimaries"/> throws
/// rather than picking one. The shape loads with ZERO errors: ValidateOnePrimarySource
/// (Loader/ValidationPasses.cs) enforces "exactly one primary" over OWN children only, and
/// MetaData.EffectiveChildren shadows an own child over a super child only on a
/// (type, name) match — so two <c>source.rdb</c> nodes with DIFFERENT explicit names at
/// two levels of an <c>extends</c> chain never collide, and both survive on the child's
/// effective <see cref="MetaObject.Sources"/>.</para>
///
/// <para>Every consumer downstream binds ONE name unconditionally, with no per-site
/// equality guard. The refusal used to live in <c>CSharpNaming</c> — so it ran only for
/// code generation, and the M:N runtime resolver silently bound the inherited PARENT's
/// relation. A refusal that depends on which consumer asked is not a refusal.</para>
///
/// <para>DIRECTION-BLIND: it compares every primary against every other, so it does not
/// matter which of them is writable nor which was declared first. Comparing against the
/// first primary WRITABLE source can only see a divergence when one of the two is
/// read-only — and, since inherited sources come first, only when the read-only one is
/// the inherited one.</para>
///
/// <para>Two primaries AGREEING on a name is not a divergence and stays legal: the
/// invariant is that an object has ONE physical name, not that it declares one source. A
/// read-only primary beside a non-primary REPLICA does not reach it either — a replica is
/// not <c>role == primary</c>.</para>
/// </summary>
public static class SourceResolution
{
    /// <summary>
    /// Every <c>@role: primary</c> source of <paramref name="obj"/>, RESOLVING through the
    /// <c>extends</c> chain (ADR-0039) — an entity inheriting its <c>source.rdb</c> from an
    /// abstract base must see it, or it would wrongly read as unpersisted.
    /// <para>Private: nothing outside this class needs the LIST. The two public entry
    /// points below are the whole surface — a caller either wants the resolved primary or
    /// only wants the refusal, and no third shape has a use.</para>
    /// </summary>
    private static IReadOnlyList<MetaSource> PrimaryRdbSources(MetaObject obj) =>
        obj.Sources().Where(s => s.Role == SOURCE_ROLE_PRIMARY).ToList();

    /// <summary>
    /// The role-scoped PRIMARY source of <paramref name="obj"/>, or null when it has none
    /// (#248: participation in the database derives from a declared primary source, never
    /// from the object subtype). Runs <see cref="RefuseDivergentPrimaries"/> first, so
    /// every caller inherits the refusal.
    /// </summary>
    public static MetaSource? PrimaryRdbSource(MetaObject obj)
    {
        var primaries = PrimaryRdbSources(obj);
        Refuse(obj, primaries);
        return primaries.Count == 0 ? null : primaries[0];
    }

    /// <summary>
    /// Refuse <paramref name="obj"/> when its <c>@role: primary</c> sources disagree on a
    /// physical name. See the class doc for the reachability analysis and for why the
    /// check is direction-blind.
    /// </summary>
    public static void RefuseDivergentPrimaries(MetaObject obj) => Refuse(obj, PrimaryRdbSources(obj));

    /// <summary>
    /// The RESOLVED address of a source — the same three parts the names artifact
    /// compares, so the shared authority and the generated artifact cannot answer
    /// "is this one object?" differently. Rendered rather than structural because every
    /// port must produce a byte-identical message from it.
    ///
    /// RAW, deliberately: an absent <c>@schema</c> is NOT folded into a dialect default.
    /// On Postgres absent and "public" address the same relation, but on SQLite/D1 they
    /// do not — the expected-schema builder rejects ANY declared schema, "public"
    /// included, while an absent one is fine. Deciding what "absent" means belongs to the
    /// caller's dialect, not to this layer.
    /// </summary>
    public static string AddressKey(MetaSource source)
    {
        string schema = source.Schema ?? "";
        string qualifier = schema.Length == 0 ? "" : $"\"{schema}\".";
        return $"{qualifier}\"{source.PhysicalName}\" ({source.EffectiveKind})";
    }

    /// <summary>
    /// The refusal itself, over a primary list the caller already has. ONE implementation
    /// and ONE copy of the message: a check written twice is a check that can disagree
    /// with itself, which is the same defect one level down from the one this class
    /// prevents (a NAME resolved twice). The message is a cross-port contract string — the
    /// other four ports carry it verbatim.
    /// </summary>
    private static void Refuse(MetaObject obj, IReadOnlyList<MetaSource> primaries)
    {
        var distinct = primaries
            .Select(AddressKey)
            .Distinct(StringComparer.Ordinal)
            .OrderBy(n => n, StringComparer.Ordinal)
            .ToList();
        if (distinct.Count <= 1) return;
        // Sorted, so the message is identical in every port regardless of source order.
        var joined = string.Join(" vs ", distinct);
        throw new InvalidOperationException(
            $"{obj.Name}: role=primary sources disagree on the object's physical address — " +
            $"{joined}. Every consumer binds ONE address. Give them a matching @kind, @schema " +
            "and physical name, or drop the extra role=primary declaration.");
    }
}
